src/services/combinePath.ts

```diff
@@ -1,26 +1,31 @@
 import { RemoveLeadingQuestionMarkFromKeys } from '@/types/params'
-import { Path, PathParamsWithParamNameExtracted, ToPath } from '@/types/path'
 import { checkDuplicateParams } from '@/utilities/checkDuplicateKeys'
+import { ParamsWithParamNameExtracted, ToWithParams, withParams, WithParams } from '@/services/withParams'
+import { StringHasValue } from '@/utilities/guards'
+
+type CombinePathString<TParent extends string | undefined, TChild extends string | undefined> =
+  StringHasValue<TParent> extends true
+    ? StringHasValue<TChild> extends true
+      ? `${TParent}${TChild}`
+      : TParent
+    : TChild
 
 export type CombinePath<
-  TParent extends Path,
-  TChild extends Path
-> = ToPath<TParent> extends { value: infer TParentPath extends string, params: infer TParentParams extends Record<string, unknown> }
-  ? ToPath<TChild> extends { value: infer TChildPath extends string, params: infer TChildParams extends Record<string, unknown> }
-    ? RemoveLeadingQuestionMarkFromKeys<TParentParams> & RemoveLeadingQuestionMarkFromKeys<TChildParams> extends PathParamsWithParamNameExtracted<`${TParentPath}${TChildPath}`>
-      ? Path<`${TParentPath}${TChildPath}`, RemoveLeadingQuestionMarkFromKeys<TParentParams> & RemoveLeadingQuestionMarkFromKeys<TChildParams>>
-      : Path<'', {}>
-    : Path<'', {}>
-  : Path<'', {}>
+  TParent extends WithParams,
+  TChild extends WithParams
+> = ToWithParams<TParent> extends { value: infer TParentPath extends string | undefined, params: infer TParentParams extends Record<string, unknown> }
+  ? ToWithParams<TChild> extends { value: infer TChildPath extends string | undefined, params: infer TChildParams extends Record<string, unknown> }
+    ? RemoveLeadingQuestionMarkFromKeys<TParentParams> & RemoveLeadingQuestionMarkFromKeys<TChildParams> extends ParamsWithParamNameExtracted<CombinePathString<TParentPath, TChildPath>>
+      ? WithParams<CombinePathString<TParentPath, TChildPath>, RemoveLeadingQuestionMarkFromKeys<TParentParams> & RemoveLeadingQuestionMarkFromKeys<TChildParams>>
+      : WithParams<undefined, {}>
+    : WithParams<undefined, {}>
+  : WithParams<undefined, {}>
```

Comment:
Do we gain something from the `undefined` vs `{}` change?

Reply:
if we just assume hash `''` (empty string) is the same as `undefined` we could remove all the `toStrings()`. See if you can remember why hash was special and expected `undefined` separately from empty string